docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+          name: "$(DATABASE_SERVICE_NAME)"
+    - 
+      kind: "ReplicationController"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        replicas: "$(REPLICA_COUNT)"
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+        template: 
+          metadata: 
+            creationTimestamp: null
+            labels: 
+              name: "$(DATABASE_SERVICE_NAME)"
+          spec: 
+            containers: 
+              - 
+                name: "mongodb"
+                image: "docker.io/centos/mongodb-26-centos7"
+                ports: 
+                  - 
+                    containerPort: 27017
+                    protocol: "TCP"
+                env: 
+                  - 
+                    name: "MONGODB_USER"
+                    value: "${MONGODB_USER}"
+                  - 
+                    name: "MONGODB_PASSWORD"
+                    value: "${MONGODB_PASSWORD}"
+                  - 
+                    name: "MONGODB_DATABASE"
+                    value: "${MONGODB_DATABASE}"
+```
+
+## Template Expansion API
+
+A server side endpoint for expanding templates is proposed.  The endpoint will inspect the `templateKind` for the template object
+being posted to the endpoint, and then use that information to determine the appropriate template expansion plugin to invoke.
+
+A template expansion plugin is expected to return a `List` of valid k8s objects, which the template processing endpoint
+will then return to the original caller.  This represents the primary contract for templates:  A template api object 
+envelope containing arbitrary content is converted into a list of k8s objects.
+
+The actual mechanism for looking up and invoking a template expansion plugin (as well as for how one is provided to k8s) is
+still being defined.  One suggestion was to essentially run a template expansion plugin as a microserver on the k8s cluster
+itself, possibly within a namespace with a name that matched the `templateKind`.  However dictating specific namespaces
+be available within the cluster for a plugin feels overly burdensome.  A map between the `templateKind` and
+the api endpoint capable of expanding it is necessary, however.
+
+Template processors could also be provided in client form to help template authors perform local testing+debugging.
+
+Nested templates should be handled by the primary endpoint, retrieving and handing each nested template off to the appropriate
+expansion plugin, and then assembling the aggregate list of objects from the top level template and any nested templates.
+
+## Open discussion points
+### Where to define parameters (closed?)
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
```

Comment:
That's actually a different issue than what I was talking about, but it's a very interesting question and quite a lot more important to discuss.

Here's how I would characterise the more interesting question:  Clearly we want a template in one language to use a template from another.  So we need a way to include templates from templates that is language-independent.  But languages also have existing importing mechanisms, that are typically limited to importing code from the same language.  Let's just consider Python, as other languages are pretty similar.  If a user wants to include a python template from another python template, they have a choice of 1) importing the python module, executing it to generate the k8s modules, then merging it in, or, 2) using a sub-template to have it merged in by k8s, in which case the two python files don't know about each other.  What is best practice?

What we're looking at is really 2 levels of modularity / encapsulation.  This feels a bit like the choice of building everything into a single binary (thus having fewer visible moving parts) or built various parts of the application as dynamically linked libraries and then these parts are re-usable.  That "fewer visilble moving parts" property will also apply to k8s if we reify the structure of the template expansion at runtime, as DM does.  If you import within Python, it will appear as a single template to k8s.  So that fundamental UX difference probably should drive the decision of one or the other.  We haven't talked about this at all yet, but if we support live migration of services from one template spec to another, the granularity may be relevant there as well.

I think if a collection of k8s objects can be considered an independent microservice that can be deployed and used by itself, it makes sense for it to be its own template.  But ultimately just like with programming, choosing how to break things up is a difficult problem that has to be handled on a case by case basis, takes a certain amount of intuition, experience and common sense.